### PR TITLE
Sync CurTime base.

### PR DESCRIPTION
**Pull request by PJB3005**
Synchronized on connect, and change of net.tickrate mid game correctly reports intended tick to client so it can make sure it's always synced.

Supersedes #2705

### CODE REVIEW

File: Robust.Shared/Timing/GameTiming.cs [hunk 1]
```diff
@@ -205,32 +194,50 @@ private TimeSpan CalcFrameTime()
...
 
             TimeSpan newTime;
 
-            if (CurTick.Value >= lastTimeTick.Value)
-              newTime = cachedTime + (TickPeriod * (CurTick.Value - lastTimeTick.Value));
+            if (atTick.Value >= lastTimeTick.Value)
+              newTime = cachedTime + (TickPeriod * (atTick.Value - lastTimeTick.Value));
             else
-              newTime = cachedTime - (TickPeriod * (lastTimeTick.Value - CurTick.Value));
+              newTime = cachedTime - (TickPeriod * (lastTimeTick.Value - atTick.Value));
```
Comment (ElectroJr) [inline Robust.Shared/Timing/GameTiming.cs]: The if else isn't even needed, is it? The two possible `newTime` expressions are equivalent.

File: Robust.Shared/Network/Messages/MsgSyncTimeBase.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Robust.Shared.Network.Messages;
+
+/// <summary>
+/// Sent server -> client to synchronize initial TimeBase on connection.
+/// Any further time synchronizations are done through the configuration system.
+/// </summary>
+internal sealed class MsgSyncTimeBase : NetMessage
+{
+    public override MsgGroups MsgGroup => MsgGroups.String;
```
Comment (ElectroJr) [inline Robust.Shared/Network/Messages/MsgSyncTimeBase.cs:13]: I have no idea what `MsgGroups` are for, but based on the doc strings this doesn't seem right?

File: Robust.Shared/Configuration/NetConfigurationManager.cs [hunk 3]
```diff
@@ -136,21 +138,35 @@ public void TickProcessMessages()
...
+                toApply.Add(msg);
 
                 _netVarsMessages.RemoveSwap(i);
                 i--;
             }
+
+            if (toApply.Count == 0)
+                return;
+
+            toApply.Sort();
```
Comment (ElectroJr) [inline Robust.Shared/Configuration/NetConfigurationManager.cs]: `MsgConVars` Doesn't appear to have `IComparable` or something else to that effect. Is this actually sorting by tick?

**Review by ElectroJr — APPROVED** [review]
Other than the sorting and minor nitpick looks all good, AFAICT this fixes the curtime issues. I haven't managed to break it so far while testing with multiple clients or high packet loss & late cvar messages.

File: Robust.Shared/Configuration/NetConfigurationManager.cs (hunk 3 above)
Comment (PJB3005) [reply Robust.Shared/Configuration/NetConfigurationManager.cs]: You're completely correct, no idea how I missed that.

File: Robust.Shared/Network/Messages/MsgSyncTimeBase.cs (hunk 2 above)
Comment (PJB3005) [reply Robust.Shared/Network/Messages/MsgSyncTimeBase.cs:13]: Basically it controls whether the messages are reliable ordered or not etc. It's a spaghetti nightmare but I can't be bothered to fix it right now so I just made it String which is ReliableOrdered.

File: Robust.Shared/Timing/GameTiming.cs (hunk 1 above)
Comment (PJB3005) [reply Robust.Shared/Timing/GameTiming.cs]: You appear to be correct in that.

**Pull request merged**